### ISSUE TITLE
Question on Model Init

### ISSUE THREAD

**Issue by XinDongol** [pytorch/torchtitan#312]
I noticed that there are two parts of implementation that are related to model initialization. 

### Instancing the model with meta tensor
https://github.com/pytorch/torchtitan/blob/f72a2a0da0bdfc394faaab9b3c0f35d0b6f5be50/train.py#L177-L181

### Doing explicit model initalization 
https://github.com/pytorch/torchtitan/blob/f72a2a0da0bdfc394faaab9b3c0f35d0b6f5be50/train.py#L209-L210

The issue is that if we do any weight initalization when instancing the module, it will ineffective becuase of the `meta tensor`. 
As a result, we have to do ***all*** initalization explicitly in the `model.init_weights()`. 

My question is why we want to instance model with `meta tensor`? 
If effencicy is not an issue, can we simply remove the `with torch.device("meta"):`

**Comment by awgu** [comment]
The advantage of meta-device init is that it is as fast as possible: the _sharded_ parameters are directly initialized on _GPU_.

Any other flow requires something more, e.g. (1) initializing unsharded parameters on GPU and then sharding or (2) initializing unsharded parameters on CPU, copying to GPU, and then the sharding. For (1), you need to insert your sharding call to be inline to your module construction, or else you will use too much GPU memory (e.g. you must construct one transformer block on GPU, shard it, construct the next transformer block, shard it, etc.). For (2), initializing parameters on CPU is slow (if you run initialization kernels), and the largest model that you can support is bottlenecked by CPU RAM size.

In some sense, the current `model.init_weights()` meta-device approach is a compromise, where we require the user to define this method to initialize all model parameters/buffers but in turn, the initialization is as fast as possible. So to answer your question, if efficiency is not an issue (and CPU RAM size is not a bottleneck), then yes, you could remove the `with torch.device("meta")` (and instead do either (1) or (2).)

**Comment by XinDongol** [comment]
Thanks for clarfying. Really helpful! 

If my understading is correct and CPU RAM is NOT a problem and if I want to do "(2)", 

```python
model = model_cls.from_model_args(model_config)
model.init_weights()
model = fully_shard(model, **fsdp_config)
model.to(device="cuda")
### training loop ###
```

Is this a correct way to do "(2)" ?

**Comment by XinDongol** [comment]
@awgu Thanks for your clarifications! 

Initalization on meta tensor is painful. For some architectures (e.g., Mamba, etc), there are some parameters need complicated pre-computed initaliztion. Assigining values to meta value is chanlleging. As a result, I would like to initalize a model before distributing it.  

I just noticed that there is a new added flag called `create_seed_checkpoint`. If my understanding is correct, it can be utilzed to avoid initalization on meta tensors.

**Comment by awgu** [comment]
I would really appreciate some pointers to the complicated initialization to learn more about it.

And yes, I think that the seed checkpoint can be used to avoid the meta device init and instead try to init on CPU.

**Comment by XinDongol** [comment]
Here are some examples. 

https://github.com/huggingface/transformers/blob/4fdf58afb72b0754da30037fc800b6044e7d9c99/src/transformers/models/rwkv/modeling_rwkv.py#L419-L439

https://github.com/huggingface/transformers/blob/4fdf58afb72b0754da30037fc800b6044e7d9c99/src/transformers/models/mamba/modeling_mamba.py#L372-L380

**Comment by awgu** [comment]
cc: @wanchaol @tianyu-l The above two pointers are good examples of real-model init methods that do not fit our current meta-device init flow. As far as I can tell, both would require some custom logic to use `DTensor` APIs to make it work (leading to some `if <using distributed>: ... else: ...`).
